svm/src/account_saver.rs

```diff
@@ -58,100 +51,87 @@ pub fn collect_accounts_to_store<'a>(
-        // to be impossible for a committable transaction to modify an
-        // invoked account if said account isn't passed to some program.
-        //
-        // Note that this assumption might not hold in the future after
-        // SIMD-0082 is implemented because we may decide to commit
-        // transactions that incorrectly attempt to invoke a fee payer or
-        // durable nonce account. If that happens, we would NOT want to use
-        // this filter because we always NEED to store those accounts even
-        // if they aren't passed to any programs (because they are mutated
-        // outside of the VM).
-        let is_storable_account = |message: &SanitizedMessage, key_index: usize| -> bool {
-            !message.is_invoked(key_index) || message.is_instruction_account(key_index)
-        };
-
-        let message = tx.message();
-        let rollback_accounts = &loaded_transaction.rollback_accounts;
-        let maybe_nonce_address = rollback_accounts.nonce().map(|account| account.address());
-
-        for (i, (address, account)) in (0..message.account_keys().len())
-            .zip(loaded_transaction.accounts.iter_mut())
-            .filter(|(i, _)| is_storable_account(message, *i))
-        {
-            if message.is_writable(i) {
-                let should_collect_account = match execution_status {
-                    Ok(()) => true,
-                    Err(_) => {
-                        let is_fee_payer = i == 0;
-                        let is_nonce_account = Some(&*address) == maybe_nonce_address;
-                        post_process_failed_tx(
-                            account,
-                            is_fee_payer,
-                            is_nonce_account,
-                            rollback_accounts,
-                            durable_nonce,
-                            lamports_per_signature,
-                        );
-
-                        is_fee_payer || is_nonce_account
-                    }
-                };
+        if executed_tx.execution_details.status.is_ok() {
+            collect_accounts_for_successful_tx(
+                &mut accounts,
+                &mut transactions,
+                tx,
+                &executed_tx.loaded_transaction.accounts,
+            );
+        } else {
+            collect_accounts_for_failed_tx(
+                &mut accounts,
+                &mut transactions,
+                tx,
+                &mut executed_tx.loaded_transaction.rollback_accounts,
+                durable_nonce,
+                lamports_per_signature,
+            );
+        }
+    }
+    (accounts, transactions)
+}
 
-                if should_collect_account {
-                    // Add to the accounts to store
-                    accounts.push((&*address, &*account));
-                    transactions.push(Some(tx));
-                }
+fn collect_accounts_for_successful_tx<'a>(
```

Comment:
nit: `collect_accounts_for_succeeded_tx`? don't count on my english sense, but i always thought that the strict counterpart of `failed` is `succeeded`, not `successful`.

Reply:
Ha this is quite interesting but actually "succeeded transaction" doesn't sound very good. Successful is the more natural sounding counterpart to "failed" when used as an adjective. Only when we are using as a verb would we say that the counterpart to "the tx failed" would be "the tx succeeded"